src/Items/ItemLilypad.h

```diff
@@ -57,51 +51,146 @@ class cItemLilypadHandler :
-					if (
-						!IsBlockWater(a_CBBlockType) &&
-						cBlockInfo::FullyOccupiesVoxel(a_CBBlockType)
-						)
-					{
-						// Can't place lilypad on air / in another block!
-						return true;
-					}
+			{	
+				if((((a_CBBlockType == E_BLOCK_STATIONARY_WATER) || (a_CBBlockType == E_BLOCK_WATER)) && (a_CBBlockMeta == 0))
+				|| (a_CBBlockType == E_BLOCK_ICE) || (a_CBBlockType == E_BLOCK_FROSTED_ICE))
+				{	
 					m_HasHitFluid = true;
 					m_Pos.Set(a_CBBlockX, a_CBBlockY, a_CBBlockZ);
+					m_BlockFace = a_CBEntryFace;
 					return true;
+				
 				}
 				return false;
 			}
 
 			Vector3i m_Pos;
+			eBlockFace m_BlockFace;
 			bool m_HasHitFluid;
-
 		};
 
 		cCallbacks Callbacks;
-		cLineBlockTracer Tracer(*a_Player->GetWorld(), Callbacks);
+		cLineBlockTracer Tracer(*a_World, Callbacks);
 		Vector3d Start(a_Player->GetEyePosition() + a_Player->GetLookVector());
 		Vector3d End(a_Player->GetEyePosition() + a_Player->GetLookVector() * 5);
 
 		Tracer.Trace(Start.x, Start.y, Start.z, End.x, End.y, End.z);
 
-		if (Callbacks.m_HasHitFluid)
+		if (!Callbacks.m_HasHitFluid)
+		{
+			return false;
+		}
+		
+		// Since we hit fluid, try placing the block on top of the fluid we found
+		// Mimick clicking on top of the fluid to do this
+		return OnPlayerPlace(*a_World, *a_Player, a_Item,
+			Callbacks.m_Pos.x, Callbacks.m_Pos.y, Callbacks.m_Pos.z, Callbacks.m_BlockFace,
+			0, 0, 0
+		);
+	}
+
+	virtual bool OnPlayerPlace(
+        cWorld & a_World, cPlayer & a_Player, const cItem & a_EquippedItem,
+        int a_BlockX, int a_BlockY, int a_BlockZ, eBlockFace a_BlockFace,
+        int a_CursorX, int a_CursorY, int a_CursorZ
+    ) override
+	{
+		Vector3i PlacementPos = GetBlockNextTo(Vector3i(a_BlockX, a_BlockY, a_BlockZ), a_BlockFace);
+		
+		// If the block where we are trying to place is a water block, try placing on top of the water block instead
+		BLOCKTYPE Placement = a_World.GetBlock(PlacementPos);
+		NIBBLETYPE PlacementMeta = a_World.GetBlockMeta(PlacementPos);
+		if(((Placement == E_BLOCK_STATIONARY_WATER) || (Placement == E_BLOCK_WATER)) && (PlacementMeta == 0))  // A water source is in the place of where we want to place the block
+		{
+			return false;
+		}	
+
+		BLOCKTYPE Surface = a_World.GetBlock(PlacementPos - Vector3i(0, 1, 0));
+		NIBBLETYPE SurfaceMeta = a_World.GetBlockMeta(PlacementPos - Vector3i(0, 1, 0));
+		if(!((((Surface == E_BLOCK_STATIONARY_WATER) || (Surface == E_BLOCK_WATER)) && (SurfaceMeta == 0)) ||  // A water source is below
+			(Surface == E_BLOCK_ICE) || (Surface == E_BLOCK_FROSTED_ICE)))                                     // Or (frosted) ice
 		{
-			a_World->SetBlock(Callbacks.m_Pos.x, Callbacks.m_Pos.y, Callbacks.m_Pos.z, E_BLOCK_LILY_PAD, 0);
-			if (!a_Player->IsGameModeCreative())
+			return false;
+		}
+
+		// Otherwise do the same exact process as cItemHandler::OnPlayerPlace, but ignore build collision
```

Comment:
Why not remove the block then call `OnPlayerPlace`?

Reply:
Remove which block? The lily pad hasn't actually been placed at this point

Reply:
Build collision implies it is overwriting a block already in that place.

Reply:
Maybe my comment as confusing, sorry, I just meant place it above the water source regardless of the build collision of water. My bad

Reply:
Okay, my main concern is that you're copy-pasting fairly complex code.